File: core-go/limiter.go
package downloadcore

import (
	"context"
	"sync"
	"time"
)

// limiter 是一个简单的令牌桶全局限速器，0 表示不限速。
type limiter struct {
	mu       sync.Mutex
	maxSpeed int64
	burst    float64
	tokens   float64
	last     time.Time
}

func newLimiter(maxSpeed int64) *limiter {
	const burst = 1 << 20 // 1 MiB
	return &limiter{
		maxSpeed: maxSpeed,
		burst:    burst,
		tokens:   burst,
		last:     time.Now(),
	}
}

// wait 阻塞直到允许读取 n 字节，或被 ctx 取消。
func (l *limiter) wait(ctx context.Context, n int64) error {
	if l == nil || l.maxSpeed <= 0 {
		return nil
	}
	for {
		l.mu.Lock()
		now := time.Now()
		elapsed := now.Sub(l.last).Seconds()
		l.last = now
		l.tokens += float64(l.maxSpeed) * elapsed
		if l.tokens > l.burst {
			l.tokens = l.burst
		}
		if l.tokens >= float64(n) {
			l.tokens -= float64(n)
			l.mu.Unlock()
			return nil
		}
		need := (float64(n) - l.tokens) / float64(l.maxSpeed)
		l.mu.Unlock()

		d := time.Duration(need * float64(time.Second))
		if d < time.Millisecond {
			d = time.Millisecond
		}
		select {
		case <-ctx.Done():
			return ctx.Err()
		case <-time.After(d):
		}
	}
}

File: core-rs/src/limiter.rs
use std::sync::Mutex;
use std::time::{Duration, Instant};

struct Inner {
    tokens: f64,
    last: Instant,
}

/// 简单的令牌桶全局限速器，max_speed = 0 表示不限速。
pub struct Limiter {
    max_speed: u64,
    burst: f64,
    inner: Mutex<Inner>,
}

impl Limiter {
    pub fn new(max_speed: u64) -> Self {
        let burst = (1 << 20) as f64; // 1 MiB
        Limiter {
            max_speed,
            burst,
            inner: Mutex::new(Inner { tokens: burst, last: Instant::now() }),
        }
    }

    /// 阻塞直到允许读取 n 字节，或被取消（返回 Err）。
    pub fn wait(&self, n: i64, canceled: &dyn Fn() -> bool) -> Result<(), ()> {
        if self.max_speed == 0 {
            return Ok(());
        }
        let n = n.max(0) as f64;
        loop {
            if canceled() {
                return Err(());
            }
            let mut g = self.inner.lock().unwrap();
            let now = Instant::now();
            let elapsed = now.duration_since(g.last).as_secs_f64();
            g.last = now;
            g.tokens += self.max_speed as f64 * elapsed;
            if g.tokens > self.burst {
                g.tokens = self.burst;
            }
            if g.tokens >= n {
                g.tokens -= n;
                return Ok(());
            }
            let need = (n - g.tokens) / self.max_speed as f64;
            drop(g);
            let mut d = Duration::from_secs_f64(need);
            if d < Duration::from_millis(1) {
                d = Duration::from_millis(1);
            }
            std::thread::sleep(d);
        }
    }
}
